### ISSUE TITLE
[Bug][RISCV] Clang produces a stable runtime result different from GCC on a reduced RV64 floating-point testcase under rva22u64/rva23u64

### ISSUE THREAD

**Issue by carlosqwqqwq** [llvm/llvm-project#192946]
### Summary

I found a stable runtime result difference between Clang and GCC on a reduced RV64 floating-point testcase.

The testcase:

- compiles successfully with both Clang and GCC,
- exits normally,
- prints a deterministic `sig:<number>` line,
- and consistently produces one signature with Clang and another with GCC.

I reproduced the difference across:

- profiles: `rva22u64`, `rva23u64`
- executors: `qemu-riscv64`, `spike`
- replay attempts: `5/5`

I am filing this in LLVM for triage because the result is stable across two executors and two profile contexts. I have not yet proven whether Clang or GCC is the incorrect side.

### LLVM version

`Ubuntu clang version 20.1.8 (++20250804090239+87f0227cb601-1~exp1~20250804210352.139)`

### Differential comparison compiler

`riscv64-linux-gnu-gcc (Debian 15.2.0-16) 15.2.0`

### Sanitizer check

I also compiled and ran the same testcase natively with host-side Clang and GCC under:

- `-fsanitize=address,undefined`

Both runs:

- compiled successfully,
- ran successfully,
- and reported no sanitizer failures.

So I do not currently have evidence that this testcase is triggering obvious UB in the source program itself.

### Reduced testcase

```c
#include <stdio.h>
#include <stdint.h>

static unsigned long fold_double(unsigned long state, double value) {
    union {
        double d;
        uint64_t u;
    } bits;
    bits.d = value;
    state ^= bits.u;
    return state * 131UL + 59UL;
}

int main(void) {
    double values[6] = {1.0, 1.75, 2.5, 3.25, 4.0, 5.5};
    double acc = 0.5 * (double)971;
    unsigned long signature = 971UL;
    for (unsigned long round = 0; round < 9; ++round) {
        for (unsigned long i = 0; i < 6; ++i) {
            double mix = values[i] * (0.5 + (double)(i + 1UL) * 0.0625);
            acc = (acc + mix) - values[(i + 2UL) % 6UL] * 0.125;
            values[i] = acc / (double)(i + 2UL) + values[(i + 1UL) % 6UL] * 0.375;
            signature = fold_double(signature + round + i, values[i]);
        }
    }
    printf("sig:%lu\n", signature);
    return 0;
}
```

### Clang reproduction

For `rva22u64`:

```bash
clang \
  --target=riscv64-linux-gnu \
  --sysroot=/usr/riscv64-linux-gnu \
  --gcc-toolchain=/usr \
  -fuse-ld=lld \
  -march=rva22u64 \
  -mabi=lp64d \
  -O2 \
  -static \
  reduced-case.c \
  -o clang-rva22u64.elf
```

Run:

```bash
qemu-riscv64 ./clang-rva22u64.elf
spike --isa=rv64imafdc_zicsr_zifencei_zicntr_zihpm_zihintpause_zba_zbb_zbs_zicbom_zicbop_zicboz_zfhmin pk ./clang-rva22u64.elf
```

For `rva23u64`:

```bash
clang \
  --target=riscv64-linux-gnu \
  --sysroot=/usr/riscv64-linux-gnu \
  --gcc-toolchain=/usr \
  -fuse-ld=lld \
  -march=rva23u64 \
  -mabi=lp64d \
  -O2 \
  -static \
  reduced-case.c \
  -o clang-rva23u64.elf
```

Run:

```bash
qemu-riscv64 ./clang-rva23u64.elf
spike --isa=rv64imafdcv_zicsr_zifencei_zicntr_zihpm_zihintpause_zba_zbb_zbs_zicbom_zicbop_zicboz_zfhmin_zfa pk ./clang-rva23u64.elf
```

### Observed behavior

The result remains stable across both executors and both profile contexts.

| Profile | Executor | Clang output | GCC output |
| --- | --- | --- | --- |
| `rva22u64` | `qemu-riscv64` | `sig:5365448365740628078` | `sig:11539865547187598679` |
| `rva22u64` | `spike` | `sig:5365448365740628078` | `sig:11539865547187598679` |
| `rva23u64` | `qemu-riscv64` | `sig:5365448365740628078` | `sig:11539865547187598679` |
| `rva23u64` | `spike` | `sig:5365448365740628078` | `sig:11539865547187598679` |

All compared runs:

- exit with code `0`
- and preserve the same compiler-partitioned result

### Expected behavior

The same well-defined reduced testcase should not produce different observable results depending only on whether it was compiled by Clang or GCC under the same RISC-V profile context.

At least one side appears to be wrong.

### Why this does not currently look executor-only

The difference stays stable across both:

- `qemu-riscv64`
- `spike`

That makes an executor-only explanation less likely.

### Why I think this is worth filing

- The testcase is already reduced and small enough to inspect directly.
- The difference is stable under replay (`5/5`).
- The difference is not tied to a single executor.
- I do not currently see evidence of obvious UB from native host-side `ASan + UBSan` checks.

### Additional details

<details>
<summary>Exact GCC reference commands used in my environment</summary>

For `rva22u64`:

```bash
riscv64-linux-gnu-gcc \
  -march=rv64imafdc_zicsr_zicntr_ziccif_ziccrse_ziccamoa_zicclsm_zic64b_za64rs_zihintpause_zba_zbb_zbs_zicbom_zicbop_zicboz_zfhmin_zkt \
  -mabi=lp64d \
  -O2 \
  -static \
  reduced-case.c \
  -o gcc-rva22u64.elf
```

For `rva23u64`:

```bash
riscv64-linux-gnu-gcc \
  -march=rv64imafdcv_zicsr_zicntr_zihpm_ziccif_ziccrse_ziccamoa_zicclsm_zic64b_za64rs_zihintpause_zba_zbb_zbs_zicbom_zicbop_zicboz_zfhmin_zkt_zvfhmin_zvbb_zvkt_zihintntl_zicond_zimop_zcmop_zcb_zfa_zawrs \
  -mabi=lp64d \
  -O2 \
  -static \
  reduced-case.c \
  -o gcc-rva23u64.elf
```

</details>

### Attachments

I am attaching:

- the reduced testcase
- the original testcase

If useful, I can also provide:

- replay summaries
- generated assembly
- additional comparison logs
- and I am happy to help reduce this further if an IR-level or codegen-level reproducer would be more useful.

**Comment by topperc** [comment]
Using clang-20 and gcc-14 on my HiFive Premier P550 board I get.

clang -O0: sig:481901421089521966
clang -O2: sig:481901421089521966
gcc -O0: sig:12165850546382151722
gcc -O2: sig:6292279728791836963

**Comment by topperc** [comment]
I get the same answer of sig:12165850546382151722 for all configurations if I use `-ffp-contract=off`

**Comment by EugeneZelenko** [comment]
Could you please try 22 or `main` branch? https://godbolt.org should be helpful. Only most recent release is maintained.

**Comment by carlosqwqqwq** [comment]
I'm sorry, it might be my mistake. I will close this issue. I will rerun it later and submit it to the community once it's ready.
